File: tests/hybrid_runtime/fixtures/fake_stdio_proxy.mjs
import fs from "node:fs";
import http from "node:http";

const [endpoint, transportFlag, transport, headerFlag, header] = process.argv.slice(2);
const token = process.env.JCODEMUNCH_HTTP_TOKEN ?? "";
if (transportFlag !== "--transport" || transport !== "http-only" ||
    headerFlag !== "--header" || header !== "Authorization: Bearer ${JCODEMUNCH_HTTP_TOKEN}" || !token) {
  process.exit(92);
}

if (process.env.JCODEMUNCH_PROXY_OBSERVATION) {
  fs.writeFileSync(process.env.JCODEMUNCH_PROXY_OBSERVATION, JSON.stringify({
    pid: process.pid,
    endpoint,
    transport,
    header_is_placeholder: header.includes("${JCODEMUNCH_HTTP_TOKEN}"),
    token_present_in_environment: Boolean(token),
    argv_contains_token: process.argv.some((value) => value.includes(token)),
  }));
}

if (process.env.JCODEMUNCH_PROXY_EXIT_IMMEDIATELY === "1") process.exit(7);

let buffered = "";
process.stdin.setEncoding("utf8");
process.stdin.on("data", (chunk) => {
  buffered += chunk;
  let newline;
  while ((newline = buffered.indexOf("\n")) >= 0) {
    const payload = buffered.slice(0, newline).trim();
    buffered = buffered.slice(newline + 1);
    if (payload) forward(payload);
  }
});

function forward(payload) {
  const target = new URL(endpoint);
  const request = http.request({
    hostname: target.hostname,
    port: Number(target.port),
    path: target.pathname,
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      Accept: "application/json, text/event-stream",
      "Content-Type": "application/json",
      "Content-Length": Buffer.byteLength(payload),
    },
  }, (response) => {
    let body = "";
    response.setEncoding("utf8");
    response.on("data", (chunk) => { body += chunk; });
    response.on("end", () => {
      const data = body.split(/\r?\n/).filter((line) => line.startsWith("data: ")).at(-1);
      if (!data) process.exit(93);
      process.stdout.write(`${data.slice(6)}\n`);
    });
  });
  request.on("error", (error) => {
    process.stderr.write(`${error.message}\n`);
    process.exit(94);
  });
  request.end(payload);
}

File: tests/hybrid_runtime/fixtures/fake_stdio_proxy.py
"""Minimal STDIO-to-HTTP proxy used to exercise the PowerShell bridge.

It intentionally validates the same argument and secret boundaries as the
pinned mcp-remote process, but leaves real mcp-remote verification to the
disposable staging and real-client acceptance gates.
"""

from __future__ import annotations

import json
import os
import sys
from pathlib import Path
from urllib.request import Request, urlopen


def main() -> int:
    arguments = sys.argv[1:]
    if len(arguments) < 6:
        return 91
    _entrypoint, endpoint, transport_flag, transport, header_flag, header = arguments[:6]
    token = os.environ.get("JCODEMUNCH_HTTP_TOKEN", "")
    if (
        transport_flag != "--transport"
        or transport != "http-only"
        or header_flag != "--header"
        or header != "Authorization: Bearer ${JCODEMUNCH_HTTP_TOKEN}"
        or not token
    ):
        return 92

    observation_path = os.environ.get("JCODEMUNCH_PROXY_OBSERVATION")
    if observation_path:
        Path(observation_path).write_text(
            json.dumps(
                {
                    "pid": os.getpid(),
                    "endpoint": endpoint,
                    "transport": transport,
                    "header_is_placeholder": "${JCODEMUNCH_HTTP_TOKEN}" in header,
                    "token_present_in_environment": bool(token),
                    "argv_contains_token": any(token in value for value in arguments),
                },
                sort_keys=True,
            ),
            encoding="utf-8",
        )

    for line in sys.stdin:
        payload = line.strip()
        if not payload:
            continue
        request = Request(
            endpoint,
            data=payload.encode("utf-8"),
            headers={
                "Authorization": f"Bearer {token}",
                "Accept": "application/json, text/event-stream",
                "Content-Type": "application/json",
            },
            method="POST",
        )
        with urlopen(request, timeout=10) as response:
            body = response.read().decode("utf-8")
        data_lines = [item[6:] for item in body.splitlines() if item.startswith("data: ")]
        if not data_lines:
            return 93
        sys.stdout.write(data_lines[-1] + "\n")
        sys.stdout.flush()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
